File: app/src/main/java/com/polok/eubmanagement/presentation/auth/signup/SignupViewModel.java
package com.polok.eubmanagement.presentation.auth.signup;

import static com.polok.eubmanagement.util.Extension.showErrorOnUi;
import android.widget.EditText;
import androidx.annotation.NonNull;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.polok.eubmanagement.base.BaseViewModel;
import com.polok.eubmanagement.base.model.OnNavigate;
import com.polok.eubmanagement.data.model.UserProfileData;
import com.polok.eubmanagement.firebase.FirebaseDataRef;
import com.polok.eubmanagement.util.FirebaseChildTag;
import com.polok.eubmanagement.util.SharedPref;

public class SignupViewModel extends BaseViewModel {
    String gender, batch, section, bloodGroup;
    String genderZero, batchZero, sectionZero, bloodGroupZero;


    void validateInputItemsAndExecuteSignup(
            EditText studentIdEt, EditText fullNameEt, EditText mobileInputEt, EditText emailInputEt, EditText passwordInputEt,
            FirebaseAuth firebaseAuth
    ) {
        if (studentIdEt.getText().toString().isEmpty()) {
            showErrorOnUi(studentIdEt, "Enter Student ID");
            return;
        }
        if (fullNameEt.getText().toString().isEmpty()) {
            showErrorOnUi(fullNameEt, "Enter Full Name");
            return;
        }
        if (mobileInputEt.getText().toString().isEmpty()) {
            showErrorOnUi(mobileInputEt, "Enter Mobile Number");
            return;
        }
        if (!android.util.Patterns.EMAIL_ADDRESS.matcher(emailInputEt.getText().toString()).matches()) {
            showErrorOnUi(emailInputEt, "Enter A Valid Email Address");
            return;
        }
        if (passwordInputEt.getText().toString().isEmpty() || passwordInputEt.getText().length() < 6) {
            showErrorOnUi(passwordInputEt, "Enter at Least 6 Digit Password");
            return;
        }
        if (gender == null || gender.isEmpty() || gender.equals(genderZero)) {
            fireMessageEvent(genderZero);
            return;
        }
        if (batch == null || batch.isEmpty() || batch.equals(batchZero)) {
            fireMessageEvent(batchZero);
            return;
        }
        if (section == null || section.isEmpty() || section.equals(sectionZero)) {
            fireMessageEvent(sectionZero);
            return;
        }
        if (bloodGroup == null || bloodGroup.isEmpty() || bloodGroup.equals(bloodGroupZero)) {
            fireMessageEvent(bloodGroupZero);
            return;
        }
        attemptSignupUserToFirebase(
                firebaseAuth,
                studentIdEt.getText().toString(),
                fullNameEt.getText().toString(),
                mobileInputEt.getText().toString(),
                emailInputEt.getText().toString(),
                passwordInputEt.getText().toString()
        );
    }

    private void attemptSignupUserToFirebase(
            FirebaseAuth firebaseAuth,
            String studentId, String fullName, String mobileInput, String emailInput, String passwordInput
    ) {
        fireLoadingEvent(true);
        firebaseAuth.createUserWithEmailAndPassword(emailInput, passwordInput).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                if (task.isSuccessful()) {
                    attemptPostUserBatchToFirebase(
                            firebaseAuth.getCurrentUser().getUid(), studentId,fullName,mobileInput,emailInput,passwordInput
                    );
                }
                else {
                    fireMessageEvent(task.getException().getLocalizedMessage());
                    fireLoadingEvent(false);
                }
            }
        });
    }

    private void attemptPostUserBatchToFirebase(
            String firebaseUid,
            String studentId, String fullName, String mobileInput, String emailInput, String passwordInput
    ) {
        //HashMap<String, String> userBatchMap = new HashMap<>();
        FirebaseDataRef.provideBatchRef().child(firebaseUid).setValue(batch).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isComplete()) try {
                    SharedPref.saveUserBatch(batch);
                } catch (Exception ignored) {
                    fireMessageEvent("Something went wrong");
                    fireLoadingEvent(false);
                } finally {
                    attemptPostUserProfileDataToFirebase(firebaseUid,studentId,fullName,mobileInput,emailInput,passwordInput);
                } else {
                    fireMessageEvent(task.getException().getLocalizedMessage());
                    fireLoadingEvent(false);
                }
            }
        });
    }

    private void attemptPostUserProfileDataToFirebase(
            String firebaseUid,
            String studentId, String fullName, String mobileInput, String emailInput, String passwordInput
    ) {
            UserProfileData userProfileData = new UserProfileData(studentId, fullName, mobileInput, emailInput, gender, section, bloodGroup, false);
            FirebaseDataRef.provideStudentRef().child(firebaseUid).child(FirebaseChildTag.PROFILE.name()).setValue(userProfileData).addOnCompleteListener(new OnCompleteListener<Void>() {
                @Override
                public void onComplete(@NonNull Task<Void> task) {
                    if (task.isComplete()) {
                        fireMessageEvent("Signup Successfully");
                        fireNavigateEvent(new OnNavigate(1));
                    } else fireMessageEvent(task.getException().getLocalizedMessage());
                    fireLoadingEvent(false);
                }
            });
    }
}

File: app/src/main/java/com/polok/eubmanagement/presentation/auth/signup/SignUpViewModel.kt
package com.polok.eubmanagement.presentation.auth.signup

import android.util.Patterns
import android.widget.EditText
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.google.firebase.auth.FirebaseAuth
import com.polok.eubmanagement.base.BaseViewModel
import com.polok.eubmanagement.firebase.FirebaseDataRef
import com.polok.eubmanagement.firebase.FirebaseDataRef.provideStudentRef
import com.polok.eubmanagement.model.UserProfileData
import com.polok.eubmanagement.util.SharedPref
import com.polok.eubmanagement.util.showErrorOnUi

class SignUpViewModel : BaseViewModel() {
    var gender: String? = null
    var batch: String? = null
    var section: String? = null
    var bloodGroup: String? = null
    var genderZero: String? = null
    var batchZero: String? = null
    var sectionZero: String? = null
    var bloodGroupZero: String? = null

    fun validateInputItemsAndExecuteSignup(
        studentIdEt: EditText,
        fullNameEt: EditText,
        mobileInputEt: EditText,
        emailInputEt: EditText,
        passwordInputEt: EditText,
        firebaseAuth: FirebaseAuth
    ) {
        if (studentIdEt.getText().toString().isEmpty()) {
            studentIdEt.showErrorOnUi("Enter Student ID")
            return
        }
        if (fullNameEt.getText().toString().isEmpty()) {
            fullNameEt.showErrorOnUi("Enter Full Name")
            return
        }
        if (mobileInputEt.getText().toString().isEmpty()) {
            mobileInputEt.showErrorOnUi("Enter Mobile Number")
            return
        }
        if (!Patterns.EMAIL_ADDRESS.matcher(emailInputEt.getText().toString()).matches()) {
            emailInputEt.showErrorOnUi("Enter A Valid Email Address")
            return
        }
        if (passwordInputEt.getText().toString()
                .isEmpty() || passwordInputEt.getText().length < 6
        ) {
            passwordInputEt.showErrorOnUi("Enter at Least 6 Digit Password")
            return
        }
        if (gender == null || gender?.isEmpty() == true || gender == genderZero) {
            fireMessageEvent(genderZero)
            return
        }
        if (batch == null || batch?.isEmpty() == true || batch == batchZero) {
            fireMessageEvent(batchZero)
            return
        }
        if (section == null || section?.isEmpty() == true || section == sectionZero) {
            fireMessageEvent(sectionZero)
            return
        }
        if (bloodGroup == null || bloodGroup?.isEmpty() == true || bloodGroup == bloodGroupZero) {
            fireMessageEvent(bloodGroupZero)
            return
        }
        attemptSignupUserToFirebase(
            firebaseAuth = firebaseAuth,
            studentId = studentIdEt.text.toString(),
            fullName = fullNameEt.text.toString(),
            mobileInput = mobileInputEt.text.toString(),
            emailInput = emailInputEt.text.toString(),
            passwordInput = passwordInputEt.text.toString()
        )
    }

    private fun attemptSignupUserToFirebase(
        firebaseAuth: FirebaseAuth,
        studentId: String,
        fullName: String,
        mobileInput: String,
        emailInput: String,
        passwordInput: String
    ) {
        fireLoadingEvent(true)
        firebaseAuth.createUserWithEmailAndPassword(emailInput, passwordInput)
            .addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    firebaseAuth.currentUser?.uid?.let {
                        attemptPostUserBatchToFirebase(
                            it, studentId, fullName, mobileInput, emailInput
                        )
                    }
                } else {
                    fireMessageEvent(task.exception!!.localizedMessage)
                    fireLoadingEvent(false)
                }
            }
    }

    private fun attemptPostUserBatchToFirebase(
        firebaseUid: String,
        studentId: String,
        fullName: String,
        mobileInput: String,
        emailInput: String
    ) {
        FirebaseDataRef.provideBatchRef()?.child(firebaseUid)?.setValue(batch)
            ?.addOnCompleteListener { task ->
                if (task.isComplete) {
                    try {
                        SharedPref.saveUserBatch(batch)
                    } catch (ignored: Exception) {
                        fireMessageEvent("Something went wrong")
                        fireLoadingEvent(false)
                    } finally {
                        attemptPostUserProfileDataToFirebase(
                            firebaseUid = firebaseUid,
                            studentId = studentId,
                            fullName = fullName,
                            mobileInput = mobileInput,
                            emailInput = emailInput
                        )
                    }
                } else {
                    fireMessageEvent(task.exception?.localizedMessage)
                    fireLoadingEvent(false)
                }
            }
    }

    private fun attemptPostUserProfileDataToFirebase(
        firebaseUid: String,
        studentId: String,
        fullName: String,
        mobileInput: String,
        emailInput: String
    ) {
        provideStudentRef()?.child(firebaseUid)?.setValue(
            UserProfileData(
                studentId = studentId,
                fullName = fullName,
                mobileNumber = mobileInput,
                email = emailInput,
                gender = gender,
                section = section,
                bloodGroup = bloodGroup,
                image = "",
                isAdmin = false
            )
        )?.addOnCompleteListener { task ->
                if (task.isComplete) {
                    fireMessageEvent("Sign up Successfully")
                    fireNavigateEvent(0, null)
                } else {
                    fireMessageEvent(task.exception?.localizedMessage)
                }
                fireLoadingEvent(false)
            }
    }

    class Factory : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return SignUpViewModel() as T
        }
    }
}